ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+
+Now let's create the top level example group. Example groups serve two purposes, they are how we organize our test file by grouping related tests together and they allow us to *describe* the behaviour we are testing within them.
+
+We define an example group using the `describe` method. In the case of the following example group we are creating an example group for all the tests related to the `TodoList` Class.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+end
+~~~
+
+Example groups can be nested within each other. This allows us to be more granular with grouping related tests together. We will see how that works by creating a nested example group where we can put all the tests for the `#name` method on the `TodoList` class.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+	describe "name" do
+
+  end
+end
+~~~
+
+The describe method takes an argument that can either be a class name or a string. Typically you will pass the name of the class you are testing to the outer most example group and use strings for the nested example groups.
+
+We will get into why you'd want to use a class name for some example groups and strings for others later on. For now, the most important thing to note is that the argument is used to provide a label that describes the behaviour we will test within the example group.
+
+In our case the the top level example group describes all the tests within it will be for the `TodoList` class.
+
+The nested example group describes all tests within it will be for the `name` method on the `TodoList` class.
+
+Structuring your test files and providing descriptive labels like this makes it easy for you to read the test file and determine what is being tested within it.
+
+The describe method also takes a block ( `do..end`). As we have already seen, nested example groups can go within the block. But we can also put individual examples within the block. Or in other words, our actual tests.
+
+#### It blocks
+
+`it` blocks are used to define the individual examples within our example groups. Much like the describe method,  `it` is also a method that accepts an argument that allows us to express what we expect to happen in that individual test, for example `it "returns the users name"`.
+
+However unlike the `describe` method, the argument **must** always be a string and we cannot nest anything within an `it` block. This is because `it` blocks are reserved for our actual test code.
+
+Let's create a few `it` blocks which we can use to test the `name` method of the todo list:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+  describe "name" do
+    it "returns the name with 0 tasks remaining" do
+    end
+
+    it "returns the name with 1 task remaining" do
+    end
+
+    it "returns the name with 10 tasks remaining" do
```

Comment:
Another very minor nitpick. This test is replaced with 2 tasks instead of 10 tasks on line 476. This is not a bit deal, but to be consistent, I suggest changing this line to 2 tasks (and also on line 302).

Reply:
Great spot 😅